rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/geo/TbGpsMultiGeofencingActionNode.java

```diff
@@ -0,0 +1,162 @@
+import org.thingsboard.rule.engine.api.TbNodeException;
+import org.thingsboard.rule.engine.util.EntitiesRelatedEntityIdAsyncLoader;
+import org.thingsboard.server.common.data.AttributeScope;
+import org.thingsboard.server.common.data.id.EntityId;
+import org.thingsboard.server.common.data.kv.AttributeKvEntry;
+import org.thingsboard.server.common.data.msg.TbMsgType;
+import org.thingsboard.server.common.data.plugin.ComponentType;
+import org.thingsboard.server.common.msg.TbMsg;
+import org.thingsboard.server.common.msg.TbMsgMetaData;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import java.util.stream.Collectors;
+
+import static org.thingsboard.common.util.DonAsynchron.withCallback;
+
+@Slf4j
+@RuleNode(type = ComponentType.ACTION,
+        name = "gps multi geofencing events",
+        configClazz = TbGpsMultiGeofencingActionNodeConfiguration.class,
+        relationTypes = {"Success", "Entered", "Left", "Inside", "Outside"},
+        nodeDescription = "Produces incoming messages using GPS based geofencing",
+        nodeDetails = "Extracts latitude and longitude parameters from incoming message and returns different events " +
+                "based on configuration parameters")
+public class TbGpsMultiGeofencingActionNode extends AbstractGeofencingNode<TbGpsMultiGeofencingActionNodeConfiguration> {
+
+    private GeofencingProcessor geofencingProcessor;
+
+    @Override
+    public void init(TbContext ctx, TbNodeConfiguration configuration) throws TbNodeException {
+        super.init(ctx, configuration);
+        geofencingProcessor = new GeofencingProcessor(ctx, this.config);
+    }
+
+    @Override
+    public void onMsg(TbContext ctx, TbMsg msg) throws ExecutionException, InterruptedException, TbNodeException {
+        ListenableFuture<List<EntityId>> matchedZonesFuture = findMatchedZones(ctx, msg);
+
+        ListenableFuture<GeofenceResponse> geofenceResponseFuture = Futures.transformAsync(matchedZonesFuture, matchedZones -> geofencingProcessor.process(msg.getMetaDataTs(), ctx.getSelfId(), msg.getOriginator(), matchedZones), ctx.getDbCallbackExecutor());
+
+        withCallback(geofenceResponseFuture, geofenceResponse -> {
+            processGeofenceResponse(ctx, msg, msg.getOriginator(), geofenceResponse);
+            ctx.tellSuccess(msg);
+        }, t -> {
+            log.error("Failed to process geofencing", t);
+            ctx.tellFailure(msg, t);
+        }, MoreExecutors.directExecutor());
+    }
+
+    private void processGeofenceResponse(TbContext ctx, TbMsg originalMsg, EntityId originatorId, GeofenceResponse geofenceResponse) {
+        TbMsgMetaData metaData = originalMsg.getMetaData().copy();
+        metaData.putValue("originatorId", originatorId.toString());
+
+        processGeofenceEvents(ctx, originalMsg, metaData, geofenceResponse.getEnteredGeofences(), "Entered");
```

Comment:
Please reuse GpsGeofencingEvents class for node connections.

```suggestion
        processGeofenceEvents(ctx, originalMsg, metaData, geofenceResponse.getEnteredGeofences(), ENTERED);
```